### PR TITLE
Feature/drodriguez/initial setup and first tests

### PR DESCRIPTION
**Pull request by diegohrlr**
## Description
Project initial setup and first POM and tests
- Added cypress framework allowing environment and browsers parameters
- Added POM skeleton with base_page, authentication pages, shared components (header+footer)
- Added first tests for authentication (Login, Sign up and Sign out)
- Added data files (urls, users)
- Added vscode settings to get a auto-format code on save to prevent alignment issues
- Added github action pipelines running on cross-browser and all devices

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x ] Documentation update

## Checklist

- [x ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: cypress/e2e/authentication/login_validation_tests.cy.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+    loginPage.pageIsDisplayed()
+  })
+
+  it('Submit login form with empty email address', () => {
+    loginPage.fillLoginFormAndSubmit('{backspace}', 'InvalidTest1234')
+    // Verifying user stuck on same page
+    loginPage.pageIsDisplayed()
+  })
+
+  it.only('Submit login form with empty password', () => {
```
Comment (ulises-jeremias) [inline cypress/e2e/authentication/login_validation_tests.cy.js:40]: **Fix Incomplete Test**

```javascript
// Current (incomplete):
it.only('Submit login form with empty password', () => {
    loginPage.fillLoginFormAndSubmit(userData.email, '{backspace}')
})

// Suggested improvement:
it('Submit login form with empty password', () => {
    loginPage.fillLoginFormAndSubmit(userData.email, '{backspace}')
    loginPage.pageIsDisplayed() // Verify user remains on login page
    // Add validation for form validation messages if they exist
})
```

File: cypress/support/pages/login_page.js [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+    doSignUpAndContinue(name, email) {
+        this.pageIsDisplayed()
+        this.fillSignUpFormAndSubmit(name, email)
+        return new SignUpPage()
+    }
+
+    verifyInvalidLoginDataAndMessage(email, pass) {
+        this.fillLoginFormAndSubmit(email, pass)
+        this.errorMessage.should('be.visible')
```
Comment (ulises-jeremias) [inline cypress/support/pages/login_page.js:65]: **Improve Wait Strategies**

```javascript
// Instead of just:
this.errorMessage.should('be.visible')

// Use:
this.errorMessage.should('be.visible')
  .and('contain.text', this.errorMessageCopy)
  .and('have.css', 'color', 'rgb(255, 0, 0)') // if error styling matters
```

**Review by ulises-jeremias — COMMENTED** [review]
Hey @diegohrlr ! 👋

Thank you so much for this solid submission! 🙌 It reflects strong foundational skills in test automation and demonstrates a clear understanding of modern testing practices with Cypress. Your Page Object Model implementation and multi-device testing approach show excellent architectural thinking.

At NaNLABS, our goal is that everyone walks away with **something valuable to explore or reflect on**, so we'll provide some constructive feedback and suggestions for growth. These insights come from a place of wanting to help you reach the next level in your automation journey 🚀

---

## 🌟 Highlights

Here are some of the things we absolutely loved:

* ✅ **Excellent Page Object Model implementation** — clean separation of concerns and proper inheritance with BasePage
* ✅ **Smart multi-device testing strategy** — comprehensive coverage across mobile, tablet, and desktop viewports
* ✅ **Professional CI/CD setup** — well-configured GitHub Actions with matrix strategy for cross-browser testing
* ✅ **Dynamic test data generation** — great use of Faker.js for realistic and varied test scenarios
* ✅ **Modular architecture** — clear separation between pages, utilities, fixtures, and shared components
* ✅ **Environment management** — proper handling of multiple environments (dev/stg/prd)
* ✅ **Self-awareness in code comments** — acknowledging technical debt shows growth mindset
* ✅ **Reporting integration** — Mochawesome setup for professional test reporting

---

## 🔧 Areas for Improvement

While your foundation is solid, here are some key areas that would take your automation skills to the next level:

### **Test Completeness & Quality**

* 🚨 **Remove `.only` modifier** in `login_validation_tests.cy.js` line 40 - this prevents other tests from running
* ✅ **Complete the incomplete test** for empty password validation with proper assertions
* 🧪 **Add more edge cases** like special characters, SQL injection attempts, and boundary value testing
* 📝 **Improve assertions** to be more specific about expected behaviors

### **Code Quality & Maintainability**

* 📁 **Extract hardcoded values** to constants file (countries list, default password, error messages)
* 🏷️ **Improve method naming** - `doSignInAndContinue` should be `doLoginAndContinue` for clarity
* 📖 **Add JSDoc comments** to document method parameters and return types
* 🔄 **Remove code duplication** in URL data files (same file exists in two locations)

### **Wait Strategies & Reliability**

* ⏱️ **Implement explicit waits** instead of relying solely on implicit waits
* 🔍 **Add proper loading state handling** for dynamic content
* 🛡️ **Improve error handling** with retry mechanisms for flaky elements

---

## 💡 Suggestions (Growth Opportunities)

These are exploratory opportunities to enhance your automation expertise:

* 🎯 **Add API integration** for test setup/teardown - create users via API, test UI, cleanup via API
* 🔍 **Implement better selector strategy** using `data-testid` attributes for more robust tests
* 📊 **Add test data management** with proper constants and configuration files
* 🚦 **Expand test coverage** to include more user flows beyond authentication
* ⚙️ **Handle network failures** and timeout scenarios for more robust tests
* 🔄 **Add custom Cypress commands** for commonly used operations
* 📈 **Implement test metrics** and coverage reporting
* 🛠️ **Add visual regression testing** for UI consistency validation

---

## 📝 Specific Code Improvements

### **Fix Incomplete Test**

```javascript
// Current (incomplete):
it.only('Submit login form with empty password', () => {
    loginPage.fillLoginFormAndSubmit(userData.email, '{backspace}')
})

// Suggested improvement:
it('Submit login form with empty password', () => {
    loginPage.fillLoginFormAndSubmit(userData.email, '{backspace}')
    loginPage.pageIsDisplayed() // Verify user remains on login page
    // Add validation for form validation messages if they exist
})
```

### **Create Constants File**

```javascript
// cypress/support/constants/index.js
export const USER_ROLES = {
  MR: 'Mr',
  MRS: 'Mrs'
}

export const COUNTRIES = [
  'India', 'United States', 'Canada', 'Australia', 
  'Israel', 'New Zealand', 'Singapore'
]

export const DEFAULT_PASSWORD = 'ChallengePass1$'

export const ERROR_MESSAGES = {
  INVALID_LOGIN: 'Your email or password is incorrect!'
}
```

### **Improve Wait Strategies**

```javascript
// Instead of just:
this.errorMessage.should('be.visible')

// Use:
this.errorMessage.should('be.visible')
  .and('contain.text', this.errorMessageCopy)
  .and('have.css', 'color', 'rgb(255, 0, 0)') // if error styling matters
```

---

## 📚 README Enhancements

Your README provides good setup instructions but could benefit from:

* **Architecture overview** explaining the Page Object Model approach
* **Test coverage details** listing what flows are automated
* **Grammar improvements** ("Cypress was used" instead of "It was used Cypress")
* **Troubleshooting section** for common issues
* **Expected results** section showing what successful test runs look like

---

## ✅ Final Thoughts

This is a **solid foundation** that demonstrates real potential in test automation! 🌟 Your architectural decisions show strong technical thinking, and your self-awareness about areas for improvement is exactly the mindset we value at NaNLABS.

With focus on completing tests, improving code quality, and adding API integration, you'll have a production-ready automation framework. The foundation you've built here shows you understand the core principles - now it's about refining the execution.

Keep up the great work and continue building on this solid foundation! 👏👏
